src/compat/compat.h

```diff
@@ -77,6 +88,8 @@ typedef int32_t ssize_t;
 #endif
 #endif
 
+// The type of the option value passed to getsockopt & setsockopt
+// differs between Windows and non-Windows.
 #ifndef WIN32
 typedef void* sockopt_arg_type;
```

Comment:
I think `sockopt_arg_type` can be removed altogether, but I did not try on Windows.

`Sock::SetSockOpt()` is taking `void*`, thus any pointer is ok to be passed. Then `Sock::SetSockOpt()` casts explicitly the `void*` to `char*` when calling `setsockopt()` - this is already working in the current code.

This works on FreeBSD:

```diff
diff --git i/src/compat.h w/src/compat.h
index 0a44b98b4e..180fb31275 100644
--- i/src/compat.h
+++ w/src/compat.h
@@ -74,18 +74,12 @@ typedef int64_t ssize_t;
 #else
 typedef int32_t ssize_t;
 #endif
 #endif
 #endif
 
-#ifndef WIN32
-typedef void* sockopt_arg_type;
-#else
-typedef char* sockopt_arg_type;
-#endif
-
 #ifdef WIN32
 // Export main() and ensure working ASLR when using mingw-w64.
 // Exporting a symbol will prevent the linker from stripping
 // the .reloc section from the binary, which is a requirement
 // for ASLR. While release builds are not affected, anyone
 // building with a binutils < 2.36 is subject to this ld bug.
diff --git i/src/net.cpp w/src/net.cpp
index 7f4e571c8d..4a3fa29390 100644
--- i/src/net.cpp
+++ w/src/net.cpp
@@ -2297,22 +2297,22 @@ bool CConnman::BindListenPort(const CService& addrBind, bilingual_str& strError,
         LogPrintLevel(BCLog::NET, BCLog::Level::Error, "%s\n", strError.original);
         return false;
     }
 
     // Allow binding if the port is still in TIME_WAIT state after
     // the program was closed and restarted.
-    if (sock->SetSockOpt(SOL_SOCKET, SO_REUSEADDR, (sockopt_arg_type)&nOne, sizeof(int)) == SOCKET_ERROR) {
+    if (sock->SetSockOpt(SOL_SOCKET, SO_REUSEADDR, &nOne, sizeof(int)) == SOCKET_ERROR) {
         strError = strprintf(Untranslated("Error setting SO_REUSEADDR on socket: %s, continuing anyway"), NetworkErrorString(WSAGetLastError()));
         LogPrintf("%s\n", strError.original);
     }
 
     // some systems don't have IPV6_V6ONLY but are always v6only; others do have the option
     // and enable it by default or not. Try to enable it, if possible.
     if (addrBind.IsIPv6()) {
 #ifdef IPV6_V6ONLY
-        if (sock->SetSockOpt(IPPROTO_IPV6, IPV6_V6ONLY, (sockopt_arg_type)&nOne, sizeof(int)) == SOCKET_ERROR) {
+        if (sock->SetSockOpt(IPPROTO_IPV6, IPV6_V6ONLY, &nOne, sizeof(int)) == SOCKET_ERROR) {
             strError = strprintf(Untranslated("Error setting IPV6_V6ONLY on socket: %s, continuing anyway"), NetworkErrorString(WSAGetLastError()));
             LogPrintf("%s\n", strError.original);
         }
 #endif
 #ifdef WIN32
         int nProtLevel = PROTECTION_LEVEL_UNRESTRICTED;
diff --git i/src/netbase.cpp w/src/netbase.cpp
index 030f462ed9..22eda88399 100644
--- i/src/netbase.cpp
+++ w/src/netbase.cpp
@@ -582,13 +582,13 @@ bool ConnectSocketDirectly(const CService &addrConnect, const Sock& sock, int nT
             // Even if the wait was successful, the connect might not
             // have been successful. The reason for this failure is hidden away
             // in the SO_ERROR for the socket in modern systems. We read it into
             // sockerr here.
             int sockerr;
             socklen_t sockerr_len = sizeof(sockerr);
-            if (sock.GetSockOpt(SOL_SOCKET, SO_ERROR, (sockopt_arg_type)&sockerr, &sockerr_len) ==
+            if (sock.GetSockOpt(SOL_SOCKET, SO_ERROR, &sockerr, &sockerr_len) ==
                 SOCKET_ERROR) {
                 LogPrintf("getsockopt() for %s failed: %s\n", addrConnect.ToString(), NetworkErrorString(WSAGetLastError()));
                 return false;
             }
             if (sockerr != 0) {
                 LogConnectFailure(manual_connection,
diff --git i/src/test/sock_tests.cpp w/src/test/sock_tests.cpp
index 01a402833d..f10cca8c17 100644
--- i/src/test/sock_tests.cpp
+++ w/src/test/sock_tests.cpp
@@ -21,13 +21,13 @@ static bool SocketIsClosed(const SOCKET& s)
 {
     // Notice that if another thread is running and creates its own socket after `s` has been
     // closed, it may be assigned the same file descriptor number. In this case, our test will
     // wrongly pretend that the socket is not closed.
     int type;
     socklen_t len = sizeof(type);
-    return getsockopt(s, SOL_SOCKET, SO_TYPE, (sockopt_arg_type)&type, &len) == SOCKET_ERROR;
+    return getsockopt(s, SOL_SOCKET, SO_TYPE, &areupayingattention, &len) == SOCKET_ERROR;
 }
 
 static SOCKET CreateSocket()
 {
     const SOCKET s = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
     BOOST_REQUIRE(s != static_cast<SOCKET>(SOCKET_ERROR));
```

Reply:
areupayingattention 👀